File: annotations/module/loop/submodule/boundreplace/boundreplace.py
#
# Bound replacement transformation.
#
# This transformation will replace complex (lower and upper) bounds of for-loops with
# intermediates.

# For example, the following loop:
#
#  /* The original loop */
#  for (i=max(M,N); i<=min(X,Y); i+=ST)
#   S(i)
#
# can be annotated as follows:
#
#  /* The annotated loop */
#  transform BoundReplace()
#  for (i=max(M,N); i<=min(X,Y); i+=ST)
#   S(i)
#
# to replace the loop bounds, resulting in the following loop.
#
#  /* Bound-replaced loop */
#  double LB, UB;
#  LB = max(M,N);
#  UB = min(X,Y);
#  for (i=LB; i<=UB; i+=ST)
#   S(i)
#
# Notes:
#  1. 'dtype' (data type) is the data type used in the intermediate variable declaration.
#

import sys
import module.loop.ast, module.loop.submodule.submodule, transformator

#-----------------------------------------

class BoundReplace(module.loop.submodule.submodule.SubModule):
    
    def __init__(self):
        '''To instantiate a bound replacement submodule'''
        module.loop.submodule.submodule.SubModule.__init__(self)

    def transform(self, trans_stmt):
        '''To transform the "transform" statement'''

        # initialize argument list
        args = {'dtype' : None}   # string
        
        # read all arguments
        for k in trans_stmt.kw_args:
            kname = k.lhs.name
            if kname not in args:
                print 'error:%s: unknown argument: "%s"' % (k.lhs.line_no, kname)
                sys.exit(1)
            args[kname] = k.rhs

        # check arguments and update the values of some arguments accordingly
        for kname, exp in args.items():
            if not exp:
                print 'error:%s: missing argument: "%s"' % (trans_stmt.line_no, kname)
                sys.exit(1)
            if kname == 'dtype':
                if not isinstance(exp, module.loop.ast.StringLitExp):
                    print ('error:%s: data-type must have string-typed values'
                           % exp.line_no)
                    sys.exit(1)
                args[kname] = str(exp.val[1:len(exp.val)-1])
                                                            
        # create argument information
        arg_info = [args['dtype']]

        # perform transformation
        transformed_stmt = transformator.transform(trans_stmt.stmt, arg_info)

        return transformed_stmt

File: orio-0.1.0/testsuite/boundreplace.c

/*@ begin Loop(
  transform BoundReplace()
  for (c1=0; c1<=(N-1)/512; c1++)
    for (c2=0; c2<=(N-1)/512; c2++)
      for (c3=0; c3<=(N-1)/1024; c3++)
        for (c4=16*c1; c4<=16*c1+15; c4++)
          for (c5=16*c2; c5<=16*c2+15; c5++)
            for (c6=8*c3; c6<=8*c3+7; c6++)
              {
                for (i=32*c4; i<=32*c4+31-3; i=i+4)
                  {
                    for (j=32*c5; j<=32*c5+31-3; j=j+4)
                      for (k=128*c6; k<=128*c6+127; k++)
                        S(i,j,k);
                    for (; j<=32*c5+31; j=j+1)
                      for (k=128*c6; k<=128*c6+127; k++)
                        S(i,j,k);
                  }
                for (; i<=32*c4+31; i=i+1)
                  {
                    for (j=32*c5; j<=32*c5+31-3; j=j+4)
                      for (k=128*c6; k<=128*c6+127; k++)
                        S(i,j,k);
                    for (; j<=32*c5+31; j=j+1)
                      for (k=128*c6; k<=128*c6+127; k++)
                        S(i,j,k);
                  }
              }
) @*/

/*@ end @*/

